### PR TITLE
Better errors alsa backend

### PR DESCRIPTION
**Pull request by JasonLG1979**
* Create and use meaningful `AlsaError`'s with `thiserror`.
* Remove all usage of `unwrap()` and `expect()` from `alsa.rs`.
* Handle all errors and never `panic!` in `alsa.rs`.
* Remove the use of `std::ffi::CString` by using `HintIter::new_str` instead of `HintIter::new`.
* Consider errors in `sink.start`, `sink.stop` and `sink.write` in `player.rs` as fatal and `exit(1)`.

### CODE REVIEW

**Comment by roderickvd** [comment]
The use of `unwrap()` and `expect()` is pervasive throughout `librespot` to the point one may think this is the intended failure mode. I can't find where, but a few months ago some of us agreed that ideally an application should never panic, but handle errors (even fatal ones) graciously. I think that's your intention here also, even if it's just one stepping stone to that end goal.

I think it's great you're picking up on Rust. And yes I think we should handle this a little differently. Generally I would choose to return meaningful error types, then handle any potential errors not in the backend, but in the `unwrap()` caller upstream (the player). This sets up error-handling that will work for all backends.

This has become a lot easier with the introduction of the [`thiserror`](https://docs.rs/thiserror/1.0.25/thiserror/) crate to `librespot` a while ago.

To see how you can set up generic error handling, you can look into the `decoder` part of the `playback` crate. That handles different error types from `VorbisEncoder` and `PassthroughEncoder` as a generic `AudioError`. You could equally use that logic for the various backends.

**Comment by JasonLG1979** [comment]
@roderickvd Thanks. like I said this is very much a wip. The problem is that from what I can tell is that alsa errors are pretty much garbage in the alsa bindings. I'll have to dig a little deeper.

I will checkout `thiserror`.

**Comment by roderickvd** [comment]
You can still throw your own meaningful error types like you do here, but then catch them in the player and exit there. What I'm trying to say is: no need to depend on the Alsa errors.

**Comment by JasonLG1979** [comment]
> You can still throw your own meaningful error types like you do here, but then catch them in the player and exit there. What I'm trying to say is: no need to depend on the Alsa errors.

I will see what I can come up with.

**Comment by JasonLG1979** [comment]
@roderickvd is that more to your liking? I also got rid of all the `unwrap`'s and `expect`'s with the exception of the ones in `list_outputs`, I didn't really think that they mattered that much since it `exit(0)`'s right after that anyway.

Now I have another problem. Since nothing panics ever now in `alsa.rs` when an error does have `librespot` is more then happy to just keep on trucking throwing infinite errors. How do you want me to deal with this? At some point we have to `exit(1)` somewhere I would think?

**Comment by JasonLG1979** [comment]
the infinite loop of error messages is ` player Could not start audio:` to `player Could not write audio`. I would think that not being able to start the sink would be considered fatal and at some point not being able to write to the sink would also be fatal?

**Comment by roderickvd** [comment]
> Now I have another problem. Since nothing panics ever now in `alsa.rs` when an error does happen `librespot` is more then happy to just keep on trucking throwing infinite errors. How do you want me to deal with this? At some point we have to `exit(1)` somewhere I would think?

You now need to handle the errors upstream in `player` that calls these backend functions.

> The infinite loop of error messages is ` player Could not start audio` to `player Could not write audio`. I would think that not being able to start the sink would be considered fatal and at some point not being able to write to the sink would also be fatal?

Agreed, you should bail out in `player`.

**Comment by JasonLG1979** [comment]
@roderickvd Their, that's all the `unwrap`'s in `alsa.rs` and as a bonus we don't need `std::ffi::CString` any more either.

**Comment by JasonLG1979** [comment]
> You now need to handle the errors upstream in player that calls these backend functions.

> Agreed, you should bail out in `player`.

Something like in `ensure_sink_running` should do it for the `Could not start audio` Error :

```rust
            match self.sink.start() {
                Ok(()) => self.sink_status = SinkStatus::Running,
                Err(err) => {
                    // Fatal error, time to bail.
                    error!("Could not start audio: {}", err);
                    exit(1);                    
                }
            }
```
Do you want me to do about the same for the `Could not write audio` error?

**Comment by JasonLG1979** [comment]
Updating the change log triggered a the CI tests. The CI tests should really ignore the change log.

**Comment by JasonLG1979** [comment]
I think that about does it @roderickvd. What do you think?

**Comment by JasonLG1979** [comment]
> I think there are a few pending review comments. Could you handle those?

What? I don't see any review comments?

> Also have you verified correct behavior on failure of the other backends?

I'll do my best. I did a quick look at the other backends and here are my observations:

* Gstreamer

The gstreamer backend has an `except` and `unwrap` at the 2 `gst::init()`'s (Honestly looking at it I'm wondering why that is called twice?)  and `write_bytes` just returns `Ok(())` no matter what. (it doesn't have `start`) So I guess those 2 could be caught and used to set a flag var so when the player tries to `write_bytes` it just return an error but that seems hacky. The gstreamer backend in general seems kinda hacky and fragile in general and doesn't seem to have a way to detect failure without reading pouring though the docs and seeing what each function returns because they don't return `Result`'s.

In any event it shouldn't be any worse off?

* JackAudio

Much cleaner/simpler than even the ALSA backend code wise. There are a couple `unwrap`'s to handle and:
```rust
            if res.is_err() {
                error!("cannot write to channel");
            }
```

Could return an error in `write` instead of `write` always returning `Ok(())` no matter what.

In any event it shouldn't be any worse off?

* Pipe

Pipe has a couple `unwarp`'s to deal with, but there's not much there to begin with.


And so on and so on....

Basically as far as I can tell my changes in `player` are going to negatively effect other backends. They'll `unwrap` and crash on error just like they did before.

I think that fixing the other backends are separate PR's?

**Comment by JasonLG1979** [comment]
PulseAudio looks good it returns errors in all the right places and doesn't have any `unwrap`'s or `except`'s.

Rodio has a couple `unwrap`'s to deal with and again always just returns `Ok(())`.

Sdl has no `unwrap`'s or `except`'s but it always returns  `Ok(())`.

Subprocess looks fine as I would assume errors returned from it would be `io:Error`'s anyway.

PortAudio strait up `panic!`'s on error, not ideal.

I can do test and fix any issues in the PulseAudio and maybe Pipe but I'm not touching the gstreamer backend with a 20 foot pole,lol!!!

**Comment by JasonLG1979** [comment]
Basically from what I can tell all the backends except PulseAudio and ALSA (after the PR) are more than happy to just always return `Ok(())' all the time and crash and burn if an error actually happens.

**Comment by roderickvd** [comment]
> What? I don't see any review comments?

e.g. https://github.com/librespot-org/librespot/pull/797/files/1d7813408f7bdb338e9bb17657c116c19d167587#diff-b9a0467929ae624ffade8358ab828a98c4ecb4dab80c15c26ba004a0e01a4f20

GitHub is a bit weird in that it doesn't show any dedicated web notifications for review comments, only mentions in the PR conversation and a notification per e-mail if you've enabled that.

> > Also have you verified correct behavior on failure of the other backends?
> 
> Basically as far as I can tell my changes in `player` are not going to negatively effect other backends. They'll `unwrap` and crash on error just like they did before.
> 
> I think that fixing the other backends are separate PR's?

Right for this PR we just need to make sure we're not causing any regressions in the other backends due to the changes in `player`.

**Comment by JasonLG1979** [comment]
> GitHub is a bit weird in that it doesn't show any dedicated web notifications for review comments, only mentions in the PR conversation and a notification per e-mail if you've enabled that.

OK. I marked them as "viewed".

> Right for this PR we just need to make sure we're not causing any regressions in the other backends due to the changes in player.

I will test the PulseAudio backend. It's the only backend that actually bothers to return anything other than `Ok(())` and not explode in the event of an error. The other backends currently could never trigger the `exit(1)` at least not cleanly if ever?

**Comment by JasonLG1979** [comment]
OK. PulseAudio works mostly as expected but `panic!`'s with a not implemented message if I try to throw `F64` at it, not great but not a regression. It would be great if it didn't `panic!` but that's really not my fault or problem. The code in player has no effect on that. I think a call should be sent out to the authors of these backends they need to be told that `panic!`ing is no longer an option and they need to return proper errors. Basically come fix your sh*t,lol!!!

**Comment by JasonLG1979** [comment]
Rodio `panic!`'s in the same way, with a not implemented message, if you try to give it a incompatible format. Again, not my fault and not my problem. The quick and dirty fix for that would be to just kick out a warning in the event of a incompatible format and change the format to the nearest compatible one. So for example in PulseAudio just make `F64` = `F32` and tell the user something like "PulseAudio does not support format F64, using F32".

**Comment by JasonLG1979** [comment]
> It looks as if `sink.write` failing is now a fatal error where it wasn't before. I really don't care either way but that should be explicit in the changelog as a material difference.

I can list that as breaking although like I have noted it really always has been unofficially fatal in that if a backend fails it just explodes.

**Comment by kingosticks** [comment]
Previously what happened depended on the backend, now it explicitly exits. This is different.

**Comment by JasonLG1979** [comment]
> Previously what happened depended on the backend, now it explicitly exits. This is different.

Different in that it's not possible to exit gracefully as apposed to sh*tting the bed,lol!!!

**Comment by JasonLG1979** [comment]
To be clear none of the backends actually handle errors completely. They all potentially `panic!`.

**Comment by JasonLG1979** [comment]
With the exception of ALSA in this PR that is...

**Comment by kingosticks** [comment]
This is a library and not everyone uses all the pieces here. You have changed what happens in one of the pieces which should be documented.

**Comment by JasonLG1979** [comment]
> This is a library and not everyone uses all the pieces here. You have changed what happens in one of the pieces which should be documented.

I'm not disagreeing with that and I will document that. But the only thing that changed outside of the alsa backend that would affect anyone is in `player` do people use `player` without a backend?

The return of `write` was already `io::Result<()>` if you're not returning an error on failure but instead `panic!`ing you're doing it wrong anyway already and your backend will crash anyway. This doesn't change that. It just allows for a graceful exit in the event that you do have a unrecoverable error and return an `io::Error`.

**Comment by JasonLG1979** [comment]
IMHO all calls to the sink should be considered fatal on error. So in `start` `stop` and `write` the backend needs to handle it's business and if it can't it needs to return an error to indicate an unrecoverable state.

**Comment by JasonLG1979** [comment]
@roderickvd I really think that `open` should also be able to return an error as it's where most of the backends do their setup currently there's no way that I can see to make that fail gracefully?

**Comment by JasonLG1979** [comment]
@kingosticks does https://github.com/librespot-org/librespot/pull/797/commits/4bb30c81b02fdd725c8873809748f4ac3e0d1c4d make you happy.

**Comment by JasonLG1979** [comment]
Again with the change log triggering CI. @roderickvd or @Johannesd3 can we fix that? It seems like a waste of time and resources for `*.md`'s to trigger a build test.

**Comment by roderickvd** [comment]
> Rodio `panic!`'s in the same way, with a not implemented message, if you try to give it a incompatible format. Again, not my fault and not my problem. The quick and dirty fix for that would be to just kick out a warning in the event of a incompatible format and change the format to the nearest compatible one. So for example in PulseAudio just make `F64` = `F32` and tell the user something like "PulseAudio does not support format F64, using F32".

For floats I'd be OK with this. But indeed not the scope of this PR. Good enough that you've checked them all out and verified they're still OK (or at least "as OK" as they were before).

> @roderickvd I really think that `open` should also be able to return an error as it's where most of the backends do their setup. Currently there's no way that I can see to make that fail gracefully?

It'd be a breaking change to change the return type and warrant a separate PR, but yes I agree. Many crates accessing the underlying sound system will also provide a `Result` so that's just nicely cascading any exceptions down (or up, as you will).

> Again with the change log triggering CI. @roderickvd or @Johannesd3 can we fix that? It seems like a waste of time and resources for `*.md`'s to trigger a build test.

Will have to take a look at it later (or take a stab at it if you will).

I'm still not very fond of the `impl fmt::Display for AudioFormat` because it already derives `Debug` and you can pretty-print with `{:?}` instead of `{}`. Can you handle the review comments I posted earlier? (also looking into the necessity for doing `._to_string()` instead of passing `&str`)

**Comment by JasonLG1979** [comment]
> For floats I'd be OK with this. But indeed not the scope of this PR. Good enough that you've checked them all out and verified they're still OK (or at least "as OK" as they were before).

Yep, they aren't pretty,lol!!!

> It'd be a breaking change to change the return type and warrant a separate PR, but yes I agree. Many crates accessing the underlying sound system will also provide a Result so that's just nicely cascading any exceptions down (or up, as you will).

I figured so. Just a thought.

> I'm still not very fond of the impl fmt::Display for AudioFormat because it already derives Debug and you can pretty-print with {:?} instead of {}

`thiserror` would not take it otherwise as a raw AudioFormat it complained that Display was not implemented so I implemented it. I can nix it and just do `format.to_string()` if you like?

> Can you handle the review comments I posted earlier?

I thought I did? Everything in your link shows as viewed? 

> also looking into the necessity for doing ._to_string() instead of passing &str

`thiserror` won't take a `str` or a `&str` without jumping though a bunch of lifetime issues. `to_string()` was much easier because it will take a `String` type just fine. If you know how to sort out the lifetime issues hook a brother up. I messed with it for a while and couldn't get it to compile. Every suggestion the compiler had just lead to more different errors.

**Comment by JasonLG1979** [comment]
thiserror would not take it otherwise as a raw AudioFormat it complained that Display was not implemented so I implemented it. I can nix it and just do format.to_string() if you like?

Nevermind I figured that part out. I will nix display.

**Comment by JasonLG1979** [comment]
@roderickvd that takes care of `impl fmt::Display for AudioFormat` for you. And unless you know how the lifetime secret combo that allows me to pass a `str` or `&str` to `thiserror` I think this is about done?

**Comment by roderickvd** [comment]
Thanks for nixing `impl Display`. For `String` vs. `&str` I think you're right so unless someone else jumps in, I'm OK with that.

To handle review comments you can mark them as "Resolved" for everyone to see, marking them as "Viewed" is just for you and does not imply "Resolved".

I'll do final sanity checking later on my system and get this in, thanks.

**Comment by JasonLG1979** [comment]
I don't see any review comments?

**Comment by JasonLG1979** [comment]
@roderickvd if you have commented on the code and started the review process that comment will be "pending" with a little yellow "pending" badge by the emoji icon in the comment header. For me to see those reviews you must go back to the main PR thread and make another comment. Otherwise any comments that are pending are only visible to the author of those comments.

**Comment by JasonLG1979** [comment]
Other than that I'm pretty sure this is done. Feel free to squash and merge as one commit at your leisure with the message being my initial PR comment. As I'm pretty sure that covers all the changes and the intermediate commits don't offer much benefit and would just be noise IMHO.

**Comment by roderickvd** [comment]
> @roderickvd if you have commented on the code and started the review process that comment will be "pending" with a little yellow "pending" badge by the emoji icon in the comment header. For me to see those reviews you must go back to the main PR comment thread (this one) and make another comment about your code comments. Otherwise any comments that are pending are only visible to the author of those comments.

🤦‍♂️ not the first time I forgot to press that button.

Looks good to me from a code review, again I'll do a final sanity run tonight and get this in.

**Comment by JasonLG1979** [comment]
> not the first time I forgot to press that button.

No worries the the reason I know about that is that I did the same thing the other day,lol!!!

**Comment by roderickvd** [comment]
I forgot to check before merging but your PR introduced quite some clippy warnings. Can you please run `cargo clippy` and open a new PR to solve these warnings?

**Comment by JasonLG1979** [comment]
> I forgot to check before merging but your PR introduced quite some clippy warnings. Can you please run `cargo clippy --features alsa-backend` and open a new PR to solve these warnings?

Sure.

Maybe put that in [CONTRIBUTING.md](https://github.com/librespot-org/librespot/blob/master/CONTRIBUTING.md)?

**Comment by roderickvd** [comment]
Yeah we should and even fail in CI if you ask me.

**Comment by JasonLG1979** [comment]
> open a new PR to solve these warnings?

Done
